ecommerce/android/Getflix/app/src/main/java/com/example/getflix/ui/viewmodels/CategoriesViewModel.kt

```diff
@@ -24,24 +24,62 @@ class CategoriesViewModel: ViewModel() {
     val products: LiveData<List<PModel>>?
         get() = _products
 
+
     var categories = arrayListOf<CategoryModel>()
 
     private var job: Job? = null
     private val exceptionHandler = CoroutineExceptionHandler { _, throwable ->
         println("Error ${throwable.localizedMessage}")
     }
 
-    fun getProducts() {
+    fun getProducts(num: Int) {
         val retrofit = Retrofit.Builder().baseUrl(MainActivity.StaticData.BASE_URL)
             .addConverterFactory(GsonConverterFactory.create())
             .build()
             .create(ProductsAPI::class.java)
         job = CoroutineScope(Dispatchers.IO).launch {
-            val response = retrofit.getProducts()
+            val response = retrofit.getProducts(num)
+            withContext(Dispatchers.Main + exceptionHandler) {
+                if (response.isSuccessful) {
+                    response.body().let { it ->
+                        _products.value = it
+                        println(_products.value.toString())
+                    }
+                }
+            }
+        }
+    }
+
+    fun getProduct(num: Int) {
+        val retrofit = Retrofit.Builder().baseUrl(MainActivity.StaticData.BASE_URL)
+                .addConverterFactory(GsonConverterFactory.create())
+                .build()
+                .create(ProductsAPI::class.java)
+        job = CoroutineScope(Dispatchers.IO).launch {
+            val response = retrofit.getProduct(num)
+            withContext(Dispatchers.Main + exceptionHandler) {
+                if (response.isSuccessful) {
+                    response.body().let { it ->
+                        _products.value = it
+                        println(_products.value.toString())
+                    }
+                }
+            }
+        }
+    }
+
+    fun getUserCartProducts(id: Int) {
+        val retrofit = Retrofit.Builder().baseUrl(MainActivity.StaticData.BASE_URL)
+                .addConverterFactory(GsonConverterFactory.create())
+                .build()
+                .create(ProductsAPI::class.java)
+        job = CoroutineScope(Dispatchers.IO).launch {
+            val response = retrofit.userCartProducts(id)
```

Comment:
I think you haven't read my remarks section carefully. As I explained, I put them to this class only for testing because necessary classes are not ready(ex. subcategory). So, I didn't think whether writing 3 lines instead of 9 is more efficient or not, we can make necessary changes while moving. Our app has also other efficiency problems(your 500 lines of code in home fragment without custom view), so maybe we open a refactoring task next week and everyone makes necessary changes to her code.

Reply:
Oh, please don't take it personally. :pensive: I think the efficiency does not come from the smaller number of codes but doing it at one shot if it is possible since, you know, the creation of the object takes time. I am not concerned with whether this could be done by shorter code or not. My purpose was how we,together, could implement functions,classes etc. as effective and robust as possible

Reply:
I'm not taking it personally, but I think we should make our reviews in a way that we respect our friends' effort. I don't understand why some efficiency problems "should be resolved" and others "would be better if resolved but not a must". You can check why a custom view is much more efficient, I think it's unnecessary to discuss it here.

Reply:
used retrofit object in GetflixAPIService now.